pkg/registry/apps/daemonset/storage/storage.go

```diff
@@ -96,6 +96,6 @@ func (r *StatusREST) Get(ctx context.Context, name string, options *metav1.GetOp
 }
 
 // Update alters the status subset of an object.
-func (r *StatusREST) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc) (runtime.Object, bool, error) {
-	return r.store.Update(ctx, name, objInfo, createValidation, updateValidation)
+func (r *StatusREST) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, config rest.UpdateConfig) (runtime.Object, bool, error) {
```

Comment:
This construct concerns me.  I suspect it's being done to avoid "forcing" people to deal with compilation errors.  But when a new concept arises, the implementers need to be aware of what's happening.  A compilation error is that indication.  If we had this container object before and simply added to it, custom storages could accidentally "work" but not actually support dry-run.

@liggitt @sttts @mfojtik we've all been on the receiving end of these signatures changing over the past 3-ish years.  I'd rather break than "work" and fail weirdly at runtime.  You guys?

Reply:
+1. I think we had the discussion about the Update signature before while adding the two admission funcs. We intentionally decided for the current way, even if it looks a bit ugly to change all implementations. But admission is important enough to break compilation if something changes.

Reply:
That's a good point, but I'm not sure how practical adding a new argument would be for something like "forceAllowCreateOnUpdate". Do we just want the signature to keep growing?

Also, I think the possibility of custom storages accidentally working when a new field is added to one of the arguments already exists, in the case of DeleteOptions, GetOptions, and when we add CreateOptions, etc.

In the case of forceAllowCreateOnUpdate, the default behavior of the new field is the behavior that we already expected from the storage, and the behavior not being performed when the caller requests it isn't going to cause any problems.

Do you think it would possibly be enough to add a comment documenting that fields shouldnt be added to the UpdateConfig without considering the implications for custom storages?

Reply:
One of the things I'm about to start arguing for more generally is "we never break api machinery consumers, ever". It is just a constraint on our output that consumers get to pick and choose the rate at which they adopt api machinery features: the old code must continue not only to compile, but also to behave the same.

We don't live in that world yet, but I want to get there.

Concretely, here, what that might look like is adding a parallel UpdateWithOptions function and allow users to implement that when they're ready to start accepting the options.

Since evolving function signatures with every change doesn't seem very maintainable, we can define a way for users to declare which semantics they understand/implement. That could just be an int that increments every type we add a feature, or it could be a struct with a bool for every feature, etc. The use sets this in their code and they only change it when they want to start using the new feature. We e.g. have a field in the `config` to hold this, and the user must set it. (This is probably not a fully-baked idea, just an example of the sort of thing that might solve the problem.)

In short, I agree silent failure is not allowable, but I don't believe that breaking compilation is the only (or best) way to fix that.

It does however seem we have to break compilation at least once to get things into the proper shape. We should make sure we're set up so that next time, we both don't break compilation AND users have to opt-in to the new behavior. (And we might as well use that mechanism so that users have to opt-in to supporting dry run, while we're at it.)

Reply:
> Concretely, here, what that might look like is adding a parallel UpdateWithOptions function and allow users to implement that when they're ready to start accepting the options.

Or we start versioning our interfaces including their methods:
```
type UpdaterV2 interface {
  UpdateV2(....)
}
```
Ugly, but might actually work and is more sustainable than `UpdateWithOptions` and `UpdateWithMoreOptions`.